recipes/openai_textcat.py

```diff
@@ -0,0 +1,284 @@
+    <summary>Show the prompt for OpenAI</summary>
+    <pre>{{openai.prompt}}</pre>
+  </details>
+  <details>
+    <summary>Show the response from OpenAI</summary>
+    <pre>{{openai.response}}</pre>
+  </details>
+</div>
+"""
+
+# Set up openai
+load_dotenv()  # take environment variables from .env.
+
+
+@dataclass
+class TextCatPromptExample(PromptExample):
+    """An example to be passed into an OpenAI TextCat prompt."""
+
+    text: str
+    answer: str
+    reason: str
+
+    @classmethod
+    def from_prodigy(cls, example: Dict, labels: Iterable[str]) -> "PromptExample":
+        """Create a prompt example from Prodigy's format."""
+        if "text" not in example:
+            raise ValueError("Cannot make PromptExample without text")
+
+        full_text = example["text"]
+        reason = example["meta"].get("reason")
+        if len(labels) == 1:
+            answer = "accept" if example.get("answer") else "reject"
+        else:
+            answer = example.get("accept")
+        return cls(text=full_text, answer=answer, reason=reason)
+
+
+def make_textcat_response_parser(labels: List[str]) -> Callable:
+    def _parse_response(text: str) -> Dict:
+        response: Dict[str, str] = {}
+        if text and any(k in text.lower() for k in ("answer", "reason")):
+            for line in text.strip().split("\n"):
+                if line and ":" in line:
+                    key, value = line.split(":", 1)
+                    response[key.strip().lower()] = value.strip()
+        else:
+            response = {"answer": "", "reason": ""}
+
+        example = _fmt_binary(response) if len(labels) == 1 else _fmt_multi(response)
+        return example
+
+    def _fmt_binary(response: Dict[str, str]) -> Dict:
+        """Parse binary TextCat where the 'answer' key means it's a positive class."""
+        return {
+            "answer": response["answer"].lower() == "accept",
+            "label": labels[0],
+            "meta": {
+                "answer": response["answer"].upper(),
+                "reason": response["reason"],
+            },
+        }
+
+    def _fmt_multi(response: Dict[str, str]) -> Dict:
+        """Parse multilabel TextCat where the 'accept' key is a list of positive labels."""
+        return {
+            "options": [{"id": label, "text": label} for label in labels],
+            "answer": "accept",
+            "meta": {"reason": response.get("reason", "")},
+            "accept": list(
+                filter(
+                    None,
+                    [normalize_label(s.strip()) for s in response["answer"].split(",")],
+                )
+            ),
+        }
+
+    return _parse_response
+
+
+@prodigy.recipe(
+    # fmt: off
+    "textcat.openai.correct",
+    input_path=("Path to jsonl data to annotate", "positional", None, Path),
+    prompt_path=("Path to jinja2 prompt template", "positional", None, Path),
+    labels=("Labels (comma delimited)", "option", "L", lambda s: s.split(",")),
```

Comment:
I wonder ... can't we automatically read the `label` from the fetched examples? 

Here's the workflow I was trying. I first downloaded some news examples. 

```
python -m prodigy textcat.openai.fetch mini-news.jsonl templates/textcat_prompt.jinja2 news-out.jsonl --labels sports -F recipes/openai_textcat.py
```

They look like this: 

```
{
  "text": "Quest For The Thanksgiving Soda GASTONIA, NC -- Last year, the Jones Soda Co. rolled out, to much fanfare, the noxious-sounding product Turkey  amp; Gravy Soda. It flew off shelves faster than free antacids on the day after Thanksgiving, and ",
  "label": "sports",
  "openai": {
    "prompt": "\n\nFrom the text below, determine whether or not it contains a sports. If it is\na sports, answer \"accept.\" If it is not a sports, answer \"reject.\"\n\n\nYour answer should only be in the following format:\n\nanswer: <string>\nreason: <string>\n\nText:\n\"\"\"\nQuest For The Thanksgiving Soda GASTONIA, NC -- Last year, the Jones Soda Co. rolled out, to much fanfare, the noxious-sounding product Turkey  amp; Gravy Soda. It flew off shelves faster than free antacids on the day after Thanksgiving, and \n\"\"\"\n",
    "response": "\nAnswer: Reject\nReason: The text does not mention any sports."
  },
  "tokens": [...],
  "meta": {
    "answer": "REJECT",
    "reason": "The text does not mention any sports."
  },
  "answer": false,
  "_input_hash": 1874369427,
  "_task_hash": 677606533
}
```

Notice how there's a `label` key in there? It might be nice (and less error prone) if I didn't have to manually pass this label when calling `textcat.openai.correct`.

```
python -m prodigy textcat.openai.correct dataname news-out.jsonl templates/textcat_prompt.jinja2 -F recipes/openai_textcat.py      

✘ textcat.teach requires at least one --label
```

There may be good reasons not to do this though, so feel free to discuss this.